xds/internal/server/rds_handler_test.go

```diff
@@ -46,33 +46,48 @@ const (
 	route3 = "route3"
 )
 
-// setupForRDSHandlerTests performs the following setup actions:
+// xdsSetupFoTests performs the following setup actions:
 //   - spins up an xDS management server
 //   - creates an xDS client with a bootstrap configuration pointing to the above
 //     management server
 //
 // Returns the following:
 // - a reference to the management server
 // - nodeID to use when pushing resources to the management server
-// - a channel to read resource names received by the management server
+// - a channel to read lds resource names received by the management server
+// - a channel to read rds resource names received by the management server
 // - an xDS client to pass to the rdsHandler under test
-func setupForRDSHandlerTests(t *testing.T) (*e2e.ManagementServer, string, chan []string, xdsclient.XDSClient) {
-	namesCh := make(chan []string, 1)
+func xdsSetupFoTests(t *testing.T) (*e2e.ManagementServer, string, chan []string, chan []string, xdsclient.XDSClient) {
+	t.Helper()
+
+	ldsNamesCh := make(chan []string, 1)
+	rdsNamesCh := make(chan []string, 1)
 
 	// Setup the management server to push the requested route configuration
 	// resource names on to a channel for the test to inspect.
 	mgmtServer, nodeID, bootstrapContents, _, cleanup := e2e.SetupManagementServer(t, e2e.ManagementServerOptions{
 		OnStreamRequest: func(_ int64, req *v3discoverypb.DiscoveryRequest) error {
-			if req.GetTypeUrl() != version.V3RouteConfigURL {
-				return fmt.Errorf("unexpected resources %v of type %q requested", req.GetResourceNames(), req.GetTypeUrl())
-			}
-			select {
-			case <-namesCh:
-			default:
-			}
-			select {
-			case namesCh <- req.GetResourceNames():
+			switch req.GetTypeUrl() {
+			case version.V3ListenerURL:
+				select {
+				case <-ldsNamesCh:
+				default:
+				}
+				select {
+				case ldsNamesCh <- req.GetResourceNames():
+				default:
+				}
```

Comment:
Selective sends and selective receives feel scary? I know you explained this in other pr, but why do you need this selectivity? Also, if you receive from the channel, why do you immediately selectively send on it, will it even be possible for another concurrent send to occur in between?